rfcs/0007-evm-support.md

```diff
@@ -0,0 +1,741 @@
+virtual machine architectures. This will also provide EVM engineers with a good balance between gas
+efficiency and safety.
+
+Currently, Solidity is the most popular language, with Vyper, Fe, and Huff being distant
+competitors. Solidity's object-oriented model, messy library support, and lack of abstract data
+types makes for a clumsy developer experience, while Vyper does not support code-sharing at all and
+makes optimization and safety tradeoffs that make it unsuitable for many modern use cases. Sway's
+type system, library support, and in-line assembly support make it a good language for the EVM.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+An EVM Backend for the sway compiler means the compiled output will be valid EVM instructions. This
+also introduces changes to the Application Binary Interface (ABI) generation to match that of the
+EVM standards defined [here](https://docs.soliditylang.org/en/v0.8.17/abi-spec.html).
+
+An EVM Sway contract may look as follows.
+
+```rust
+// imports all evm types
+use std::evm_prelude::*;
+
+// Storage definition
+storage {
+    // EVM Type: Uint256
+    value: Uint256,
+}
+
+// Event definition
+struct ValueSet {
+    // Indexed<T> is a type to be used only in Event structs
+    old: Indexed<Uint256>,
+    new: Indexed<Uint256>,
+}
+
+// ABI definition
+abi ValueHandler {
+    #[storage(read, write)]
+    fn set(new: Uint256);
+    #[storage(read)]
+    fn value() -> Uint256;
+}
+
+// Contract definition
+impl ValueHandler for Contract {
+    #[storage(read, write)]
+    fn set(new: Uint256) {
+        let old: Uint256 = storage.value;
+        storage.value = new;
+
+        log(ValueSet { old, new });
+    }
+
+    #[storage(read)]
+    fn value() -> Uint256 {
+        storage.value
+    }
+}
+```
+
+Compare this to a functionally equivalent Solidity contract.
+
+```sol
+// ABI definition
+interface IValueHandler {
+    function set(uint256 newValue) external;
+    function value() external view returns (uint256);
+}
+
+// Contract definition
+contract ValueHandler is IValueHandler {
+    event ValueSet(uin256 indexed oldValue, uint256 indexed newValue);
+
+    // Storage layout is implicit and follows C3 linearization with
+    // lower-order alignment and is packed by each type's max size.
+    // The `public` keyword implicitly creates a view function.
+    uint256 public value;
+
+    function set(uint256 newValue) external {
+        uint256 oldValue = value;
+        value = newValue;
+
+        emit ValueSet(oldValue, newValue);
+    }
+}
+```
+
+EVM Code generation can be enabled by a new field in the `Forc.toml` file.
+
+```toml
+[project]
+authors = ["user"]
+entry = "main.sw"
+license = "Apache-2.0"
+name = "example"
+target = "eofv1"
+```
+
+Development, Testing, and Deployment can be done through Hardhat or Foundry using a simple plugin
+for each.
```

Comment:
Hmm this makes me think if we are planning to support sway unit tests for EVM backend cc @mitchmindtree 

Currently `forc-test` is coupled with `FuelVM`, if we are going to be supporting tests for EVM as well we might need to think about it a little.

Reply:
Yeah I think to start we'd want to find a nice embeddable EVM interpreter - I'm guessing there might be a lib from foundry we could plug-in for this?

Reply:
We already have an EVM testing framework set up for the E2Es which might work for unit tests too.  Not sure, @tritao looked at it initially.

Reply:
Nice 👌, it will be a good reference. Thanks!